### ISSUE TITLE
countDistinct returns null, instead of 0 for null types

### ISSUE THREAD

**Issue by mighTY94** [ClickHouse/ClickHouse#2494]
Steps to reproduce:

**1. Create table with the following structure:**

```
xxxx :) CREATE TABLE nullableCountTest ( type UInt32, subType Nullable(String)) ENGINE=Log

CREATE TABLE nullableCountTest
(
    type UInt32,
    subType Nullable(String)
)
ENGINE = Log

Ok.

0 rows in set. Elapsed: 0.002 sec.
```

**2. Insert some data, similar to what I've inserted:**

```
xxxx :) INSERT INTO nullableCountTest VALUES (1, 'myType'), (2, 'myType2'), (3, 'myType3'), (4, 'myType'), (5, Null)

INSERT INTO default.nullableCountTest VALUES

Ok.

5 rows in set. Elapsed: 0.003 sec.
```

**3. Make a distinctCount query**

```
xxxx :) SELECT subType, countDistinct(subType) FROM nullableCountTest GROUP BY subType;

SELECT
    subType,
    countDistinct(subType)
FROM nullableCountTest
GROUP BY subType

┌─subType─┬─uniqExact(subType)─┐
│ myType3 │                  1 │
│ myType  │                  1 │
│ myType2 │                  1 │
│ \N      │                 \N │
└─────────┴────────────────────┘

4 rows in set. Elapsed: 0.003 sec.
```

It might be a bug, or expected behaviour, but what if I want to process the result: 

```
xxxx :) SELECT if(countDistinct(subType) IS NULL, '0 Items', concat(CAST(countDistinct(subType) AS Char), ' Items')) FROM default.nullableCountTest GROUP BY subType;

SELECT if(isNull(countDistinct(subType)), '0 Items', concat(CAST(countDistinct(subType) AS Char), ' Items'))
FROM nullableCountTest
GROUP BY subType

Received exception from server (version 1.1.54383):
Code: 349. DB::Exception: Received from localhost:9000, 127.0.0.1. DB::Exception: Cannot convert NULL value to non-Nullable type.

0 rows in set. Elapsed: 0.113 sec.
```

It doesn't allow conversion of null to another type. Also, it looks like both the expressions in the IF are evaluated regardless if the condition is met. Then the condition decides which one to use. I might be wrong, but the ifnull would be working then..

_the simple count function works as expected, though:_

```
xxxx :) SELECT subType, count(subType) FROM nullableCountTest GROUP BY subType;

SELECT
    subType,
    count(subType)
FROM nullableCountTest
GROUP BY subType

┌─subType─┬─count(subType)─┐
│ myType3 │              1 │
│ myType  │              2 │
│ myType2 │              1 │
│ \N      │              0 │
└─────────┴────────────────┘

4 rows in set. Elapsed: 0.003 sec.

```

**Comment by KochetovNicolai** [comment]
In your case you can use `assumeNotNull` function:
```
:)  SELECT if(countDistinct(subType) IS NULL, '0 Items', concat(CAST(assumeNotNull(countDistinct(subType)) AS Char), ' Items')) FROM default.nullableCountTest GROUP BY subType

SELECT if(isNull(countDistinct(subType)), '0 Items', concat(CAST(assumeNotNull(countDistinct(subType)) AS Char), ' Items'))
FROM default.nullableCountTest 
GROUP BY subType

┌─if(isNull(uniqExact(subType)), \'0 Items\', concat(CAST(assumeNotNull(uniqExact(subType)), \'Char\'), \' Items\'))─┐
│ 1 Items                                                                                                            │
│ 1 Items                                                                                                            │
│ 1 Items                                                                                                            │
│ 0 Items                                                                                                            │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

```

**Comment by chris-baynes** [comment]
I believe this is a bug, the simplest reproduction case is just:

```
:) select count(distinct null)

SELECT countDistinct(NULL)

┌─uniqExact(NULL)─┐
│ \N              │
└─────────────────┘
```

`count` & `countDistinct` should always return an Int >= 0. This is currently an issue for us as it breaks an integration tool that assumes the result will not be null, which is valid assumption for all other databases we connect to.

**Comment by alexey-milovidov** [comment]
@Akazz We can introduce a property of aggregate functions like `returnDefaultWhenOnlyNull`. If this property is set, we will not wrap the result type in Nullable and will return default value when nothing was aggregated.

**Comment by filimonov** [comment]
Similar #7336 #6741

Maybe covered (or not?) by #7331

**Comment by filimonov** [comment]
It is still there:
```
SELECT countDistinct(toNullable(number))
FROM numbers(1)
WHERE 1 = 0

┌─uniqExact(toNullable(number))─┐
│                          ᴺᵁᴸᴸ │
└───────────────────────────────┘

1 rows in set. Elapsed: 0.001 sec. 

SELECT uniqExactOrDefault(toNullable(number))
FROM numbers(1)
WHERE 1 = 0

┌─uniqExactOrDefault(toNullable(number))─┐
│                                   ᴺᵁᴸᴸ │
└────────────────────────────────────────┘

1 rows in set. Elapsed: 0.004 sec. 


SELECT uniqExactOrNull(toNullable(number))
FROM numbers(1)
WHERE 1 = 0

┌─uniqExactOrNull(toNullable(number))─┐
│                                ᴺᵁᴸᴸ │
└─────────────────────────────────────┘

1 rows in set. Elapsed: 0.004 sec. 
```

**Comment by alexey-milovidov** [comment]
Fixed by #11661
